test/utils/test-Variant.cc

```diff
@@ -0,0 +1,895 @@
+//
+//  test/utils/test-Variant.cc
+//  test-librefract
+//
+//  Created by Thomas Jandecka on 17/01/2018
+//  Copyright (c) 2018 Apiary Inc. All rights reserved.
+//
+
+#include <catch.hpp>
+
+#include "../Tracker.h"
+#include <string>
+
+#include "utils/Variant.h"
+
+struct Foo : tracked<Foo> {
+    int data_ = {};
+
+    Foo() = default;
+    Foo(const Foo&) = default;
+    Foo(Foo&&) = default;
+    Foo& operator=(const Foo&) = default;
+    Foo& operator=(Foo&&) = default;
+    ~Foo() = default;
+
+    explicit Foo(int data) : data_(data) {}
+};
+
+struct Bar : tracked<Bar> {
+    bool data_ = {};
+
+    Bar() = default;
+    Bar(const Bar&) = default;
+    Bar(Bar&&) = default;
+    Bar& operator=(const Bar&) = default;
+    Bar& operator=(Bar&&) = default;
+    ~Bar() = default;
+
+    explicit Bar(bool data) : data_(data) {}
+};
+
+struct Baz : tracked<Baz> {
+    std::string data_ = {};
+
+    Baz() = default;
+    Baz(const Baz&) = default;
+    Baz(Baz&&) = default;
+    Baz& operator=(const Baz&) = default;
+    Baz& operator=(Baz&&) = default;
+    ~Baz() = default;
+
+    explicit Baz(std::string data) : data_(std::move(data)) {}
+};
+
+using namespace drafter;
+using namespace utils;
+
+using tested = variant<Foo, Bar, Baz>;
+
+SCENARIO("variant construction", "[variant][init]")
+{
+    GIVEN("A default intitialized variant<Foo, Bar, Baz>")
+    {
+        Foo::reset_record();
+        Bar::reset_record();
+        Baz::reset_record();
+
+        tested v;
+
+        THEN("only the first variant type exists")
+        {
+            REQUIRE(Foo::instances().size() == 1);
+            REQUIRE(Bar::instances().size() == 0);
+            REQUIRE(Baz::instances().size() == 0);
+
+            THEN("it is allocated somewhere in the variant")
+            {
+                using address = const std::uint8_t*;
+
+                address v_mem_start = reinterpret_cast<address>(&v);
+                address v_mem_end = v_mem_start + sizeof(v);
+
+                address foo_mem_start = reinterpret_cast<address>(&Foo::last_instance());
+
+                REQUIRE(foo_mem_start >= v_mem_start);
+                REQUIRE(foo_mem_start < v_mem_end);
+            }
+
+            THEN("it was default constructed")
+            {
+                REQUIRE(Foo::record().default_constructor == 1);
+                REQUIRE(Foo::record().copy_constructor == 0);
+                REQUIRE(Foo::record().move_constructor == 0);
+            }
+
+            THEN("it was not copied")
+            {
+                REQUIRE(Foo::record().copy_constructor == 0);
+                REQUIRE(Foo::record().copy_assignment == 0);
+            }
+
+            THEN("it was not moved")
+            {
+                REQUIRE(Foo::record().copy_constructor == 0);
+                REQUIRE(Foo::record().copy_assignment == 0);
+            }
```

Comment:
I don't see the above 2 `THEN` anywhere else in similar tests below. Did you forget to remove these?

Reply:
👍 actually `move_contructor` and `move_assignment` should be tested there

Reply:
Shouldn't you be testing `move_assignment` and `copy_assignment` in other tests below too?

Reply:
👍 added assertions where missing